**Question 2**

**Implement a simple convolutional neural network for handwritten digit
classification using the MNIST dataset. Please implement a normal artificial
neural network, then a CNN, and compare the performance.**


In [ ]:
#Importing required libraries
import tensorflow as tf
from tensorflow.keras import datasets, layers, models
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
#Load dataset
(X_train, y_train), (X_test,y_test) = datasets.mnist.load_data()
X_train.shape

11490434/11490434 [==============================] - 0s 0us/step


(60000, 28, 28)

In [ ]:
X_test.shape

(10000, 28, 28)

In [ ]:
y_train.shape

(60000,)

In [ ]:
#Normalize the images to a number from 0 to 1. Image has 3 channels (R,G,B) and each value in the channel can range from 0 to 255.
#Hence to normalize in 0-->1 range, we need to divide it by 255

X_train = X_train / 255.0
X_test = X_test / 255.0

#Build simple artificial neural network for image classification

In [ ]:
ann = models.Sequential([
        layers.Flatten(input_shape=(28,28,1)),
        layers.Dense(3000, activation='relu'),
        layers.Dense(1000, activation='relu'),
        layers.Dense(10, activation='softmax')
    ])

ann.compile(optimizer='SGD',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

ann.fit(X_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 91s 48ms/step - loss: 0.5018 - accuracy: 0.8774
Epoch 2/5
1875/1875 [==============================] - 76s 41ms/step - loss: 0.2495 - accuracy: 0.9291
Epoch 3/5
1875/1875 [==============================] - 71s 38ms/step - loss: 0.2002 - accuracy: 0.9434
Epoch 4/5
1875/1875 [==============================] - 72s 38ms/step - loss: 0.1681 - accuracy: 0.9527
Epoch 5/5
1875/1875 [==============================] - 74s 39ms/step - loss: 0.1440 - accuracy: 0.9594


In [ ]:
from sklearn.metrics import confusion_matrix , classification_report
import numpy as np
y_pred = ann.predict(X_test)
y_pred_classes = [np.argmax(element) for element in y_pred]

print("Classification Report: \n", classification_report(y_test, y_pred_classes))

313/313 [==============================] - 6s 20ms/step
Classification Report: 
               precision    recall  f1-score   support

           0       0.97      0.99      0.98       980
           1       0.98      0.99      0.99      1135
           2       0.97      0.96      0.96      1032
           3       0.94      0.97      0.95      1010
           4       0.96      0.96      0.96       982
           5       0.95      0.95      0.95       892
           6       0.96      0.97      0.96       958
           7       0.96      0.95      0.96      1028
           8       0.98      0.92      0.95       974
           9       0.95      0.95      0.95      1009

    accuracy                           0.96     10000
   macro avg       0.96      0.96      0.96     10000
weighted avg       0.96      0.96      0.96     10000



#Now let's build a convolutional neural network to train our images

In [ ]:
cnn = models.Sequential([
    layers.Conv2D(filters=32, kernel_size=(3, 3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D((2, 2)),

    layers.Conv2D(filters=64, kernel_size=(3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),

    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

In [ ]:
cnn.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [ ]:
cnn.fit(X_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 46s 24ms/step - loss: 0.1356 - accuracy: 0.9582
Epoch 2/10
1875/1875 [==============================] - 47s 25ms/step - loss: 0.0450 - accuracy: 0.9859
Epoch 3/10
1875/1875 [==============================] - 45s 24ms/step - loss: 0.0316 - accuracy: 0.9898
Epoch 4/10
1875/1875 [==============================] - 46s 25ms/step - loss: 0.0243 - accuracy: 0.9923
Epoch 5/10
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0181 - accuracy: 0.9940
Epoch 6/10
1875/1875 [==============================] - 43s 23ms/step - loss: 0.0141 - accuracy: 0.9955
Epoch 7/10
1875/1875 [==============================] - 44s 24ms/step - loss: 0.0109 - accuracy: 0.9964
Epoch 8/10
1875/1875 [==============================] - 47s 25ms/step - loss: 0.0097 - accuracy: 0.9966
Epoch 9/10
1875/1875 [==============================] - 44s 24ms/step - loss: 0.0079 - accuracy: 0.9976
Epoch 10/10
1875/1875 [==============================] - 44s 24m

**With CNN, at the end 5 epochs, accuracy was at around 99% which is a significant improvement over ANN. CNN's are best for image classification and gives superb accuracy. Also computation is much less compared to simple ANN as maxpooling reduces the image dimensions while still preserving the features**

In [ ]:
cnn.evaluate(X_test,y_test)

313/313 [==============================] - 2s 7ms/step - loss: 0.0442 - accuracy: 0.9893


[0.04422392696142197, 0.989300012588501]

In [ ]:
y_pred = cnn.predict(X_test)
y_pred

313/313 [==============================] - 4s 11ms/step


array([[1.28992346e-12, 2.20284083e-10, 1.84227833e-09, ...,
        9.99999940e-01, 5.56800751e-11, 4.69909745e-12],
       [8.52588995e-13, 1.62887079e-10, 9.99999940e-01, ...,
        1.53511965e-14, 4.50505146e-15, 2.80538223e-15],
       [5.50631318e-10, 9.99998152e-01, 4.96451047e-10, ...,
        1.74749289e-07, 6.33934462e-07, 8.67037020e-09],
       ...,
       [3.29324149e-21, 7.06784674e-13, 4.49270535e-18, ...,
        1.41474204e-14, 1.01934317e-09, 9.07711666e-12],
       [1.11980807e-11, 1.07145625e-16, 2.65079520e-14, ...,
        5.54073420e-16, 1.84006851e-10, 1.45021474e-15],
       [3.16923085e-11, 4.14617091e-10, 8.89087581e-10, ...,
        5.01517070e-15, 7.11755543e-09, 1.22165444e-10]], dtype=float32)

In [ ]:
y_classes = [np.argmax(element) for element in y_pred]
y_classes[:5]

[7, 2, 1, 0, 4]

In [ ]:
from sklearn.metrics import confusion_matrix , classification_report
import numpy as np
y_pred = cnn.predict(X_test)
y_pred_classes = [np.argmax(element) for element in y_pred]

print("Classification Report: \n", classification_report(y_test, y_pred_classes))

313/313 [==============================] - 3s 9ms/step
Classification Report: 
               precision    recall  f1-score   support

           0       0.99      1.00      0.99       980
           1       0.99      1.00      0.99      1135
           2       0.99      0.99      0.99      1032
           3       0.98      1.00      0.99      1010
           4       1.00      0.99      0.99       982
           5       0.98      0.99      0.99       892
           6       0.99      0.98      0.99       958
           7       0.98      0.99      0.99      1028
           8       1.00      0.98      0.99       974
           9       0.99      0.99      0.99      1009

    accuracy                           0.99     10000
   macro avg       0.99      0.99      0.99     10000
weighted avg       0.99      0.99      0.99     10000



**Classification Report Metrics**

**Precision**: The ratio of correctly predicted positive observations to the total predicted positives. High precision relates to the low false positive rate.

**Recall**: The ratio of correctly predicted positive observations to all the observations in the actual class. High recall relates to the low false negative rate.

**F1-Score**: The weighted average of Precision and Recall. This score is useful when the class distribution is imbalanced.

**Support**: The number of actual occurrences of the class in the dataset.

** Comparing both ANN and CNN**

**CNN Performance:**

**Higher Accuracy:** CNN achieves 99% accuracy compared to 96% with ANN.

**Better Precision and Recall:** CNN has better precision and recall for all classes, meaning it makes fewer errors in classification.

Higher F1-Scores: CNN consistently has higher F1-scores, indicating a better balance between precision and recall.

**Implications:**

CNN models are more suited for image data as they can capture spatial hierarchies through convolutional and pooling layers.

ANN models may not perform as well on image data due to the lack of convolutional layers to extract spatial features.